pkg/apis/apps/v1alpha1/cloneset_types.go

```diff
@@ -0,0 +1,196 @@
+	// The maximum number of pods that can be unavailable during the update.
+	// Value can be an absolute number (ex: 5) or a percentage of desired pods (ex: 10%).
+	// Absolute number is calculated from percentage by rounding down.
+	// Defaults to 1.
+	MaxUnavailable *intstr.IntOrString `json:"maxUnavailable,omitempty"`
+	// Paused indicates that the CloneSet is paused.
+	// Default value is false
+	Paused bool `json:"paused,omitempty"`
+	// InPlaceUpdateStrategy contains strategies for in-place update.
+	InPlaceUpdateStrategy *CloneSetInPlaceUpdateStrategy `json:"inPlaceUpdateStrategy,omitempty"`
+}
+
+type CloneSetUpdateStrategyType string
+
+const (
+	// RecreateCloneSetUpdateStrategyType indicates that we always delete Pod and create new Pod
+	// during Pod update, which is the default behavior.
+	RecreateCloneSetUpdateStrategyType CloneSetUpdateStrategyType = "ReCreate"
+	// InPlaceIfPossibleCloneSetUpdateStrategyType indicates that we try to in-place update Pod instead of
+	// recreating Pod when possible. Currently, only image update of pod spec is allowed. Any other changes to the pod
+	// spec will fall back to ReCreate CloneSetUpdateStrategyType where pod will be recreated.
+	InPlaceIfPossibleCloneSetUpdateStrategyType CloneSetUpdateStrategyType = "InPlaceIfPossible"
+	// InPlaceOnlyCloneSetUpdateStrategyType indicates that we will in-place update Pod instead of
+	// recreating pod. Currently we only allow image update for pod spec. Any other changes to the pod spec will be
+	// rejected by kube-apiserver
+	InPlaceOnlyCloneSetUpdateStrategyType CloneSetUpdateStrategyType = "InPlaceOnly"
+)
+
+type CloneSetInPlaceUpdateStrategy struct {
+}
+
+type CloneSetUpdatePriorityTerm struct {
+	// Weight associated with matching the corresponding matchExpressions, in the range 1-100.
+	Weight int32 `json:"weight"`
+	// MatchSelector is used to select by pod's labels.
+	MatchSelector metav1.LabelSelector `json:"matchSelector"`
+}
+
+// CloneSetStatus defines the observed state of CloneSet
+type CloneSetStatus struct {
+	// ObservedGeneration is the most recent generation observed for this CloneSet. It corresponds to the
+	// CloneSet's generation, which is updated on mutation by the API Server.
+	ObservedGeneration int64 `json:"observedGeneration,omitempty"`
+
+	// Replicas is the number of Pods created by the CloneSet controller.
+	Replicas int32 `json:"replicas"`
+
+	// ReadyReplicas is the number of Pods created by the CloneSet controller that have a Ready Condition.
+	ReadyReplicas int32 `json:"readyReplicas"`
+
+	// UpdatedReplicas is the number of Pods created by the CloneSet controller from the CloneSet version
+	// indicated by updateRevision.
+	UpdatedReplicas int32 `json:"updatedReplicas"`
+
+	// UpdatedReadyReplicas is the number of Pods created by the CloneSet controller from the CloneSet version
+	// indicated by updateRevision and have a Ready Condition.
+	UpdatedReadyReplicas int32 `json:"updatedReadReplicas"`
```

Comment:
I think there is a typo here, which causes the API violation notice in api/api-rules/violation_exceptions.list, should be  `json:"updatedReadyReplicas"`

Reply:
done